conftest.py

```diff
@@ -0,0 +1,151 @@
+    "cstar/externals/**"
+]
+
+
+EXAMPLE_BLUEPRINTS = {
+    "ROMS_MARBL": {
+        # TODO this assumes you are running pytest from the root directory of the cloned repo, which is fragile
+        "base": "./examples/cstar_blueprint_roms_marbl_example.yaml",
+        "input_datasets_url_prefix": "https://github.com/CWorthy-ocean/input_datasets_roms_marbl_example/raw/main/",
+        "additional_code_url": "https://github.com/CWorthy-ocean/cstar_blueprint_roms_marbl_example.git",
+    },
+}
+
+
+@pytest.fixture
+def mock_user_input():
+    """
+    Monkeypatch which will automatically respond to any call for input.
+    
+    Use it like this:
+ 
+        ```
+        def some_test(mock_user_input):
+            with mock_user_input("yes"):
+                assert input("Enter your choice: ") == "yes"
+        ```
+    """
+    @contextmanager
+    def _mock_input(input_string):
+        original_input = builtins.input
+        def mock_input_function(_):
+            return input_string
+        builtins.input = mock_input_function
+        try:
+            yield
+        finally:
+            builtins.input = original_input
+    
+    return _mock_input
+
+
+@pytest.fixture
+def example_blueprint_as_dict() -> Callable[[str], dict]:
+    """Given the name of a pre-defined blueprint, return it as an in-memory dict."""
+
+    def _base_blueprint_dict(name: str) -> dict:
+        base_blueprint_path = EXAMPLE_BLUEPRINTS[name]["base"]
+
+        with open(base_blueprint_path, "r") as file:
+            base_blueprint_dict = yaml.safe_load(file)
+
+        return base_blueprint_dict
+
+    return _base_blueprint_dict
+
+
+# TODO should these local files be in a specified temporary directory?
+
+def change_remote_paths_to_local_paths(
+    blueprint: dict,
+    name: str,
+) -> dict:
+    """
+    Alter an in-memory blueprint to point to the locally-downloaded versions of files downloaded from remote data sources.
+
+    Parameters
+    ----------
+    blueprint
+    name
+    """
+
+    input_datasets_url_prefix: str = EXAMPLE_BLUEPRINTS[name][
+        "input_datasets_url_prefix"
+    ]
+    additional_code_url: str = EXAMPLE_BLUEPRINTS[name]["additional_code_url"]  # noqa
+
+    def is_a_location_containing_input_datasets_url(key: str, value: str) -> bool:
+        return key == "location" and input_datasets_url_prefix in value
+
+    def update_path_to_input_datasets(key: str, value: str) -> str:
+        filename = Path(value).name
+        # TODO make "local_input_files" a parameter?
+        local_filepath = Path.cwd() / "local_input_files" / filename
+        return str(local_filepath)
+
+    blueprint_with_local_input_datasets = process_nested_dict(
+        dictionary=blueprint,
+        condition=is_a_location_containing_input_datasets_url,
+        update=update_path_to_input_datasets,
+    )
+
+    return blueprint_with_local_input_datasets
+
+    # TODO
+    # blueprint_with_local_additional_code = ...
```

Comment:
My main question with this PR is what is going on with local additional code? 

I expected to have to implement this function here in order for the test to pass, but it already passes without me having implemented this function. Does that mean this test doesn't actually test the use of local additional code at all?

Reply:
Turns out the local test was passing but for the wrong reasons. The dict was not being edited correctly, and the pre-existing remote paths were being kept around, so the second test ended up basically being equivalent to the first test.

Obviously our tests should not pass if they behave unexpectedly! A better solution (as suggested by @dafyddstephenson ) might be to start with an invalid blueprint that contains placeholders of somekind, and fill it in with valid paths for both tests. Rather than starting from a valid input and changing it to another valid input. That way if the replacement step fails the input will be invalid, and the test fail.

Reply:
Thanks for noting this down - it may be worth adding assert statements for things like `{AdditionalCode,InputDataset}.source.location_type` to verify the modification has gone ahead also